miden-tx/src/host/note_builder.rs

```diff
@@ -65,14 +65,35 @@ impl OutputNoteBuilder {
                 return Err(TransactionKernelError::MalformedRecipientData(data.to_vec()));
             }
             let inputs_hash = Digest::new([data[0], data[1], data[2], data[3]]);
-            let inputs_key = NoteInputs::commitment_to_key(inputs_hash);
             let script_hash = Digest::new([data[4], data[5], data[6], data[7]]);
             let serial_num = [data[8], data[9], data[10], data[11]];
-            let input_els = adv_provider.get_mapped_values(&inputs_key);
             let script_data = adv_provider.get_mapped_values(&script_hash).unwrap_or(&[]);
 
-            let inputs = NoteInputs::new(input_els.map(|e| e.to_vec()).unwrap_or_default())
-                .map_err(TransactionKernelError::MalformedNoteInputs)?;
+            let inputs_data = adv_provider.get_mapped_values(&inputs_hash);
+            let inputs = match inputs_data {
+                None => NoteInputs::default(),
+                Some(inputs) => {
+                    if inputs.is_empty() {
+                        return Err(TransactionKernelError::MissingNoteInputs);
+                    }
+
+                    let num_inputs = u64::from(inputs[0]) as usize;
+                    if num_inputs > (inputs.len() - 1) {
+                        return Err(TransactionKernelError::MissingNoteInputs);
+                    }
```

Comment:
We use that name already [`MalformedNoteInputs`](https://github.com/0xPolygonMiden/miden-base/blob/6f1a03c9d1fdde87ffd1c4e64cfcab6ee7709102/miden-lib/src/transaction/errors.rs#L24), and this error is returned if the `NoteInputs::new` [constructor fails](https://github.com/0xPolygonMiden/miden-base/blob/e6c5293f61ccbcbe43406d382d3453b0a3447ad7/miden-tx/src/host/mod.rs#L146-L147). It seemed cleaner to keep the two separate. Do you want to change it?

Reply:
Ah yes - probably we should mix these. Should we create a different error variant then? It doesn't seem like `MissingNoteInputs` is the right variant for this error either.

Reply:
Added `TooManyElementsForNoteInputs`

Reply:
Looking at this now, shouldn't the condition be `num_inputs != (inputs.len() - 1)`? That is, if there are too few elements that's an error too - right?

Reply:
My rationale was that this would fail in the commitment check below. Changed to `!=`

Reply:
Ah, forgot about the padding. This needs to be `>`

Reply:
Makes sense! Could you add a short comment explaining this? Otherwise, looking at the code it may not be immediately clear why it is OK for the vector to have more elements than `num_inputs`.

Reply:
added a comment